qubes/mgmt.py

```diff
@@ -186,11 +234,239 @@ def vm_property_help(self, untrusted_payload):
+        # reject duplicated options
+        assert len(set(x[0] for x in untrusted_pool_config)) == \
+               len([x[0] for x in untrusted_pool_config])
+        # and convert to dict
+        untrusted_pool_config = dict(untrusted_pool_config)
+
+        assert 'name' in untrusted_pool_config
+        untrusted_pool_name = untrusted_pool_config.pop('name')
+        allowed_chars = string.ascii_letters + string.digits + '-_.'
+        assert all(c in allowed_chars for c in untrusted_pool_name)
+        pool_name = untrusted_pool_name
+        assert pool_name not in self.app.pools
+
+        driver_parameters = qubes.storage.driver_parameters(self.arg)
+        assert all(key in driver_parameters for key in untrusted_pool_config)
+
+        # option names validated, validation of option values is delegated to
+        #  extension (through events mechanism)
+        self.fire_event_for_permission(name=pool_name,
+            untrusted_pool_config=untrusted_pool_config)
+        pool_config = untrusted_pool_config
+
+        self.app.add_pool(name=pool_name, driver=self.arg, **pool_config)
+        self.app.save()
+
+    @asyncio.coroutine
+    @no_payload
+    def pool_remove(self):
+        assert self.dest.name == 'dom0'
+        assert self.arg in self.app.pools.keys()
+
+        self.fire_event_for_permission()
+
+        self.app.remove_pool(self.arg)
+        self.app.save()
+
+    @asyncio.coroutine
+    @no_payload
+    def label_list(self):
+        assert self.dest.name == 'dom0'
+        assert not self.arg
+
+        labels = self.fire_event_for_filter(self.app.labels.values())
+
+        return ''.join('{}\n'.format(label.name) for label in labels)
+
+    @asyncio.coroutine
+    @no_payload
+    def label_get(self):
+        assert self.dest.name == 'dom0'
+
+        try:
+            label = self.app.get_label(self.arg)
+        except KeyError:
+            raise qubes.exc.QubesValueError
+
+        self.fire_event_for_permission(label=label)
+
+        return label.color
+
+    @asyncio.coroutine
+    def label_create(self, untrusted_payload):
+        assert self.dest.name == 'dom0'
+
+        # don't confuse label name with label index
+        assert not self.arg.isdigit()
+        allowed_chars = string.ascii_letters + string.digits + '-_.'
+        assert all(c in allowed_chars for c in self.arg)
+        try:
+            self.app.get_label(self.arg)
+        except KeyError:
+            # ok, no such label yet
+            pass
+        else:
+            raise qubes.exc.QubesValueError('label already exists')
+
+        untrusted_payload = untrusted_payload.decode('ascii').strip()
+        assert len(untrusted_payload) == 8
+        assert untrusted_payload.startswith('0x')
+        # besides prefix, only hex digits are allowed
+        assert all(x in string.hexdigits for x in untrusted_payload[2:])
+
+        # TODO: try to avoid creating label too similar to existing one?
+        color = untrusted_payload
```

Comment:
Sure. In that case, we can file a bug to remember, paste the snippet and remove TODO.